File: src/builtins/macro.rs
use crate::builtins::*;
use crate::fundamentals::*;
use std::rc::Rc;

#[derive(Debug, Clone)]
pub struct MacroParameter(pub Name);

#[derive(Clone)]
pub struct DefineMacroParameterFn(
    pub Rc<dyn Fn(Value, &mut Environment, &ProgramStack) -> Result<(MacroParameter, Value)>>,
);

impl DefineMacroParameterFn {
    pub fn new(
        define_parameter: impl Fn(Value, &mut Environment, &ProgramStack) -> Result<(MacroParameter, Value)>
            + 'static,
    ) -> DefineMacroParameterFn {
        DefineMacroParameterFn(Rc::new(define_parameter))
    }
}

simple_trait! {
    name: macro_parameter,
    type: DefineMacroParameterFn,
    label: "Macro-Parameter",
}

#[derive(Clone)]
pub struct MacroExpandFn(
    pub Rc<dyn Fn(MacroParameter, Value, &mut Environment, &ProgramStack) -> Result>,
);

impl MacroExpandFn {
    pub fn new(
        expand: impl Fn(MacroParameter, Value, &mut Environment, &ProgramStack) -> Result + 'static,
    ) -> MacroExpandFn {
        MacroExpandFn(Rc::new(expand))
    }
}

simple_trait! {
    name: macro_expand,
    type: MacroExpandFn,
    label: "Macro-Expand",
}

impl Value {
    pub fn macro_expand(
        &self,
        parameter: MacroParameter,
        replacement: Value,
        env: &mut Environment,
        stack: &ProgramStack,
    ) -> Result {
        let stack = stack.add(&format!(
            "Expanding '{}', replacing '{}' with '{}'",
            self.format(env, stack),
            parameter.0 .0,
            replacement.format(env, stack)
        ));

        match self.get_trait_if_present(TraitID::macro_expand, env, &stack)? {
            Some(expand) => expand.0(parameter, replacement, env, &stack),
            None => Ok(self.clone()),
        }
    }
}

#[derive(Clone)]
pub struct Macro {
    pub define_parameter: DefineMacroParameterFn,
    pub value_to_expand: Value,
}

simple_trait! {
    name: r#macro,
    type: Macro,
    label: "Macro",
}

pub(crate) fn init(env: &mut Environment) {
    // Macro ::= Function
    env.add_conformance(Conformance::new(
        TraitID::function,
        TraitID::r#macro.validation(),
        |r#macro, _, _| {
            let r#macro = r#macro.clone();

            Ok(Function::new(move |input, env, stack| {
                let (parameter, replacement) = r#macro.define_parameter.0(input, env, stack)?;

                r#macro
                    .value_to_expand
                    .macro_expand(parameter, replacement, env, stack)?
                    .evaluate(env, stack)
            }))
        },
    ));

    // Macro ::= Text
    env.add_conformance(Conformance::new(
        TraitID::text,
        TraitID::r#macro.validation(),
        |_, _, _| Ok(Text(String::from("<macro>"))),
    ));
}

File: Sources/Builtins/Macro.swift
public typealias MacroParameter = String

public typealias DefineMacroParameterFunc = (Value, inout Environment, ProgramStack) throws -> (
    MacroParameter, Value
)

extension TraitID where T == DefineMacroParameterFunc {
    public static let macroParameter = TraitID(debugLabel: "Macro-Parameter")
}

extension Trait where T == DefineMacroParameterFunc {
    public static func macroParameter(_ value: @escaping DefineMacroParameterFunc) -> Self {
        Trait(id: .macroParameter) { _, _ in value }
    }
}

public typealias MacroExpandFunc = (MacroParameter, Value, inout Environment, ProgramStack) throws
    -> Value

extension TraitID where T == MacroExpandFunc {
    public static let macroExpand = TraitID(debugLabel: "Macro-Expand")
}

extension Trait where T == MacroExpandFunc {
    public static func macroExpand(_ value: @escaping MacroExpandFunc) -> Self {
        Trait(id: .macroExpand) { _, _ in value }
    }
}

extension Value {
    public func macroExpand(
        parameter: MacroParameter,
        replacement: Value,
        _ env: inout Environment,
        _ stack: ProgramStack
    ) throws -> Value {
        let stack =
            stack
            .add(
                "Expanding '\(self.format(&env, stack))', replacing '\(parameter)' with '\(replacement.format(&env, stack))'"
            )

        guard let macroExpand = try self.traitIfPresent(.macroExpand, &env, stack) else {
            return self
        }

        return try macroExpand(parameter, replacement, &env, stack)
    }
}

public struct Macro {
    public var defineParameter: DefineMacroParameterFunc
    public var valueToExpand: Value

    public init(
        defineParameter: @escaping DefineMacroParameterFunc,
        valueToExpand: Value
    ) {
        self.defineParameter = defineParameter
        self.valueToExpand = valueToExpand
    }
}

extension TraitID where T == Macro {
    public static let macro = TraitID(debugLabel: "Macro")
}

extension Trait where T == Macro {
    public static func macro(_ value: Macro) -> Self {
        Trait(id: .macro) { _, _ in value }
    }
}

internal func setupMacro(_ env: inout Environment) {
    // Macro ::= Function
    env.addConformance(
        Conformance(
            derivedTraitID: .function,
            validation: TraitID.macro.validation,
            deriveTraitValue: { macro, _, _ in
                { value, env, stack in
                    let (parameter, replacement) = try macro.defineParameter(value, &env, stack)

                    return try macro.valueToExpand.macroExpand(
                        parameter: parameter,
                        replacement: replacement,
                        &env,
                        stack
                    )
                }
            }
        )
    )

    // Macro ::= Text
    env.addConformance(
        Conformance(
            derivedTraitID: .text,
            validation: TraitID.macro.validation,
            deriveTraitValue: { _, _, _ in
                Text("<macro>")
            }
        )
    )
}
